sui/contract/sources/constant_product.move

```diff
@@ -89,154 +144,201 @@ module aux::constant_product {
         assert!(reserves_x > 0 && reserves_y > 0, EPoolEmpty);
 
-        let base_amount = get_amount_out(
-            (coin::value(&coin_y) as u128),
+        let amount_out = get_amount_out(
+            (amount_in as u128),
             (reserves_y as u128),
             (reserves_x as u128),
             (self.fee_bps as u128)
         );
+        assert!(amount_out >= min_amount_x, EMinSwap);
 
-        assert!(base_amount >= min_base_amount, EMinSwap);
-
+        event::emit<Swapped<Y, X>>(Swapped { amount_in, amount_out });
         coin::join(&mut self.reserve_y, coin_y);
-        coin::take(coin::balance_mut(&mut self.reserve_x), base_amount, ctx)
+        coin::take(coin::balance_mut(&mut self.reserve_x), amount_out, ctx)
     }
 
-    entry fun mint_<X, Y, W>(
-        self: &mut Pool<X, Y, W>,
-        coin_x: Coin<X>,
-        coin_y: Coin<Y>,
-        ctx: &mut TxContext
-    ) {
-        transfer::transfer(
-            mint(self, coin_x, coin_y, 0, ctx),
-            tx_context::sender(ctx)
-        );
-    }
-
-
     /// TODO slippage
     /// 
-    /// say you tolerate 1% slippage, so adding 100 SUI and 1 USDC, you provide between
+    /// say you tolerate 1% slippage, so adding 100 COIN_X and 1 COIN_Y, you provide between
     /// 
-    /// 99 SUI / 1 USDC
-    /// 100 SUI / 0.99 USDC
-    public fun mint<X, Y, W>(
-        self: &mut Pool<X, Y, W>,
+    /// 99 COIN_X / 1 COIN_Y
+    /// 100 COIN_X / 0.99 COIN_Y
+    public fun add_liquidity<X, Y>(
+        self: &mut Pool<X, Y>,
         coin_x: Coin<X>,
         coin_y: Coin<Y>,
-        _slippage_bps: u64,  // TODO
         ctx: &mut TxContext
-    ): Coin<LP<X, Y, W>> {
+    ): Coin<LP<X, Y>> {
         assert!(coin::value(&coin_x) > 0, EInvalidArg);
         assert!(coin::value(&coin_y) > 0, EInvalidArg);
 
-        let x = coin::value(&coin_x);
-        let y = coin::value(&coin_y);
+        let amount_added_x = coin::value(&coin_x);
+        let amount_added_y = coin::value(&coin_y);
 
         // TODO burn initial tokens
         let lp_supply = balance::supply_value(&self.supply_lp);
-        let lp = if (lp_supply == 0) {
-            assert!(x > 0 && y > 0, EInvalidArg);
-            math::sqrt(x * y)
+        let amount_minted_lp = if (lp_supply == 0) {
+            assert!(amount_added_x > 0 && amount_added_y > 0, EInvalidArg);
+            math::sqrt(amount_added_x) * math::sqrt(amount_added_y)
         } else {
             let pool_x = coin::value(&self.reserve_x);
             let pool_y = coin::value(&self.reserve_y);
             // calc share based on ratio of existing deposits
-            let share_x = (x * lp_supply) / pool_x;
-            let share_y = (y * lp_supply) / pool_y;
+            // FIXME overflow (but dividing separately has integer division issue)
+            let share_x = (amount_added_x * lp_supply) / pool_x;
+            let share_y = (amount_added_y * lp_supply) / pool_y;
```

Comment:
Can probably just cast to `u128`

Reply:
right, but casting `share` back to `u64` can still overflow? i guess we don't worry about this in aptos

Reply:
Yeah well actually in sui you could deal with overflowing coin values by splitting into multiple coins since there's no coin store but seems unnecessary